cmd/otelcol/main.go

```diff
@@ -193,16 +235,42 @@ func checkMemorySettingsMiBFromEnvVar(envVar string) int {
 			log.Fatalf("Expected a number greater than 0 for %s env variable but got %s", envVar, envVarVal)
 		}
 		envVarResult = val
-	} else {
+	case memTotalSize > 0:
+		break
+	default:
 		log.Printf("Usage: %s=12345 %s=us0 %s=684 %s=1024 %s=256 %s", tokenEnvVarName, realmEnvVarName, ballastEnvVarName, memLimitMiBEnvVarName, memSpikeMiBEnvVarName, os.Args[0])
 		log.Fatalf("ERROR: Missing environment variable %s", envVar)
 	}
 	return envVarResult
 }
 
-func useMemorySettingsMiBFromEnvVar() {
-	memLimit := checkMemorySettingsMiBFromEnvVar(memLimitMiBEnvVarName)
-	memSpike := checkMemorySettingsMiBFromEnvVar(memSpikeMiBEnvVarName)
+func useMemorySettingsMiBFromEnvVar(memTotalSize int) {
+	memLimit := checkMemorySettingsMiBFromEnvVar(memLimitMiBEnvVarName, memTotalSize)
+	if memLimit == 0 {
+		if memTotalSize == 0 {
+			panic("PANIC: Both memory limit MiB and memory total size are set to zero. This should never happen.")
+		}
+		memLimitMiB := memTotalSize * defaultMemoryLimitPercentage / 100
+		if (memTotalSize - memLimitMiB) < defaultMemoryLimitMaxMiB {
+			memLimit = memLimitMiB
+		} else {
+			memLimit = (memTotalSize - defaultMemoryLimitMaxMiB)
+		}
```

Comment:
It is not clear what this calculation does and what's relationship between defaultMemoryLimitPercentage and defaultMemoryLimitMaxMiB and why both exist.

Reply:
percentage is percentage from total memory size and max is the maximum of amount of memory. I added a comment to hopefully clarify. In short, we want to set the limit to 90% of total memory reserving a maximum of 2048 of memory while for spike at 25% for a maximum of 2048 of memory.